cpuminer.go

```diff
@@ -508,3 +509,112 @@ func newCPUMiner(s *server) *CPUMiner {
 		updateHashes:      make(chan uint64),
 	}
 }
+
+// generateNBlocks generates the requested number of blocks. It is self
+// contained in that it creates block templates and attempts to solve them while
+// detecting when it is performing stale work and reacting accordingly by
+// generating a new block template.  When a block is solved, it is submitted.
+// The function returns a list of the hashes of generated blocks.
+func (m *CPUMiner) generateNBlocks(n uint32) ([]*wire.ShaHash, error) {
+	m.Lock()
+
+	// Respond with an error if there's virtually 0 chance of CPU-mining a block.
+	if !m.server.chainParams.GenerateSupported {
+		m.Unlock()
+		return nil, errors.New("No support for `generate` on the current " +
+			"network, " + m.server.chainParams.Net.String() +
+			", as it's unlikely to be possible to CPU-mine a block.")
+	}
+
+	// Respond with an error if server is already mining.
+	if m.started {
+		m.Unlock()
+		return nil, errors.New("Server is already CPU mining. Please call " +
+			"`setgenerate 0` before calling discrete `generate` commands.")
+	}
+
+	m.started = true
+
+	m.speedMonitorQuit = make(chan struct{})
+	m.wg.Add(1)
+	go m.speedMonitor()
+
+	m.Unlock()
+
+	minrLog.Tracef("Generating %d blocks", n)
+
+	i := uint32(0)
+	blockHashes := make([]*wire.ShaHash, n, n)
+
+	// Start a ticker which is used to signal checks for stale work and
+	// updates to the speed monitor.
+	ticker := time.NewTicker(time.Second * hashUpdateSecs)
+	defer ticker.Stop()
+
+	for {
+		// Wait until there is a connection to at least one other peer
+		// since there is no way to relay a found block or receive
+		// transactions to work on when there are no connected peers.
+		if m.server.ConnectedCount() == 0 {
```

Comment:
This makes since for the normal mining path, but I don't think it should be checked for the discreet `generate` path.  It is pretty common to want to create a few blocks against a single node when doing regression testing.